api/crl/crl.go

```diff
@@ -1,104 +1,90 @@
+	ca         *x509.Certificate
+	key        crypto.Signer
 }
 
-// Handle responds to requests for crl generation. It creates this crl
-// based off of the given certificate, serial numbers, and private key
-func crlHandler(w http.ResponseWriter, r *http.Request) error {
-
-	var revokedCerts []pkix.RevokedCertificate
-	var oneWeek = time.Duration(604800) * time.Second
-	var newExpiryTime = time.Now()
-
-	body, err := ioutil.ReadAll(r.Body)
+// NewHandler returns a new http.Handler that handles a revoke request.
+func NewHandler(dbAccessor certdb.Accessor, caPath string, caKeyPath string) (http.Handler, error) {
+	ca, err := ioutil.ReadFile(caPath)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	r.Body.Close()
-
-	req := &jsonCRLRequest{}
 
-	err = json.Unmarshal(body, req)
+	caKey, err := ioutil.ReadFile(caKeyPath)
 	if err != nil {
-		log.Error(err)
+		return nil, errors.Wrap(errors.CertificateError, errors.ReadFailed, err)
 	}
 
-	if req.ExpiryTime != "" {
-		expiryTime := strings.TrimSpace(req.ExpiryTime)
-		expiryInt, err := strconv.ParseInt(expiryTime, 0, 32)
-		if err != nil {
-			return err
-		}
-
-		newExpiryTime = time.Now().Add((time.Duration(expiryInt) * time.Second))
+	// Parse the PEM encoded certificate
+	issuerCert, err := helpers.ParseCertificatePEM(ca)
+	if err != nil {
+		return nil, err
 	}
 
-	if req.ExpiryTime == "" {
-		newExpiryTime = time.Now().Add(oneWeek)
+	strPassword := os.Getenv("CFSSL_CA_PK_PASSWORD")
+	password := []byte(strPassword)
+	if strPassword == "" {
+		password = nil
 	}
 
+	// Parse the key given
+	key, err := helpers.ParsePrivateKeyPEMWithPassword(caKey, password)
 	if err != nil {
-		return err
+		log.Debug("Malformed private key %v", err)
+		return nil, err
 	}
 
-	cert, err := helpers.ParseCertificatePEM([]byte(req.Certificate))
-	if err != nil {
-		log.Error("Error from ParseCertificatePEM", err)
-		return errors.NewBadRequestString("Malformed certificate")
-	}
+	return &api.HTTPHandler{
+		Handler: &Handler{
+			dbAccessor: dbAccessor,
+			ca:         issuerCert,
+			key:        key,
+		},
+		Methods: []string{"GET"},
+	}, nil
+}
 
-	for _, value := range req.SerialNumber {
-		tempBigInt := new(big.Int)
-		tempBigInt.SetString(value, 10)
-		tempCert := pkix.RevokedCertificate{
-			SerialNumber:   tempBigInt,
-			RevocationTime: time.Now(),
-		}
-		revokedCerts = append(revokedCerts, tempCert)
-	}
+// Handle responds to revocation requests. It attempts to revoke
+// a certificate with a given serial number
+func (h *Handler) Handle(w http.ResponseWriter, r *http.Request) error {
+	var newExpiryTime = "0"
 
-	key, err := helpers.ParsePrivateKeyPEM([]byte(req.PrivateKey))
+	certs, err := h.dbAccessor.GetRevokedAndUnexpiredCertificates()
 	if err != nil {
-		log.Debug("Malformed private key %v", err)
-		return errors.NewBadRequestString("Malformed Private Key")
+		return err
+	}
+
+	queryExpiryTime := r.URL.Query().Get("expiry")
+	if queryExpiryTime != "" {
+		log.Info("Requested Expiry Time")
```

Comment:
Log messages shouldn't be capitalised like this and it'd be useful to include the requested expiry, e.g. `log.Infof("requested expiry time of %s", expiry)`